src/plans/nonlinear_least_squares_plan.jl

```diff
@@ -1,132 +1,178 @@
 
-# Constructors
+# Arguments
+
+* `f` the vectorial cost function ``f: $(_math(:M)) → ℝ^m``
+* `jacobian` the Jacobian, might also be a vector of gradients of the component functions of `f`
+* `range_dimension` the number of dimensions `m` the function `f` maps into
+
+These three can also be passed as a [`AbstractVectorGradientFunction`](@ref) `vf` already.
 
-    NonlinearLeastSquaresProblem(M, F, jacF, num_components; evaluation=AllocatingEvaluation(), jacobian_tangent_basis=DefaultOrthonormalBasis())
+# Keyword arguments
+
+$(_var(:Keyword, :evaluation))
+* `function_type::`[`AbstractVectorialType`](@ref)`=`[`FunctionVectorialType`](@ref)`()`: format the
+* `jacobian_tangent_basis::AbstractBasis=DefaultOrthonormalBasis()`; shortcut to specify the basis the Jacobian matrix is build with.
+* `jacobian_type::`[`AbstractVectorialType`](@ref)`=`[`CoordinateVectorialType`](@ref)`(jacobian_tangent_basis)`: specify the format the Jacobian is given in. By default a matrix of the differential with respect to a certain basis of the tangent space.
 
 # See also
 
 [`LevenbergMarquardt`](@ref), [`LevenbergMarquardtState`](@ref)
 """
-struct NonlinearLeastSquaresObjective{E<:AbstractEvaluationType,TC,TJ,TB<:AbstractBasis} <:
-       AbstractManifoldGradientObjective{E,TC,TJ}
-    f::TC
-    jacobian!!::TJ
-    jacobian_tangent_basis::TB
-    num_components::Int
+struct NonlinearLeastSquaresObjective{
+    E<:AbstractEvaluationType,F<:AbstractVectorGradientFunction{E}
+} <: AbstractManifoldGradientObjective{E,F,F}
+    objective::F
 end
+
 function NonlinearLeastSquaresObjective(
-    f::TF,
-    jacobian_f::TJ,
-    num_components::Int;
+    f,
+    jacobian,
+    range_dimension;
     evaluation::AbstractEvaluationType=AllocatingEvaluation(),
-    jacobian_tangent_basis::TB=DefaultOrthonormalBasis(),
-) where {TF,TJ,TB<:AbstractBasis}
-    return NonlinearLeastSquaresObjective{typeof(evaluation),TF,TJ,TB}(
-        f, jacobian_f, jacobian_tangent_basis, num_components
+    jacobian_tangent_basis::AbstractBasis=DefaultOrthonormalBasis(),
+    jacobian_type::AbstractVectorialType=CoordinateVectorialType(jacobian_tangent_basis),
+    function_type::AbstractVectorialType=FunctionVectorialType(),
+    kwargs...,
+)
+    vgf = VectorGradientFunction(
+        f,
+        jacobian,
+        range_dimension;
+        evaluation=evaluation,
+        jacobian_type=jacobian_type,
+        function_type=function_type,
     )
+    return NonlinearLeastSquaresObjective(vgf; kwargs...)
 end
 
+# Cost
 function get_cost(
-    M::AbstractManifold, nlso::NonlinearLeastSquaresObjective{AllocatingEvaluation}, p
-)
-    return 1//2 * norm(nlso.f(M, p))^2
+    M::AbstractManifold,
+    nlso::NonlinearLeastSquaresObjective{
+        E,<:AbstractVectorFunction{E,<:ComponentVectorialType}
+    },
+    p;
+    kwargs...,
+) where {E<:AbstractEvaluationType}
+    v = 0.0
+    for i in 1:length(nlso.objective)
+        v += abs(get_value(M, nlso.objective, p, i))^2
```

Comment:
Since the values are real, `abs` seems to be redundant here I think?

Reply:
By the way, IMO the canonical implementation should be the component-free one for nonlinear least squares, not the one that accesses components one by one.

Reply:
probably a bit, yes, depending on how you interpret the objective/residual. 

Might be work to keep this in case we generalise to complex residuals later.

Reply:
> By the way, IMO the canonical implementation should be the component-free one for nonlinear least squares, not the one that accesses components one by one 

I do not understand what you mean, I think I did not change logic anywhere, I just wrapped f and jacobian into the vector function. The case here is where single functions are stored ( $f = (f_1, f_2, ... f_n)$) hence the component wise access. If one function is stored, this is not done.

Reply:
> Might be work to keep this in case we generalise to complex residuals later.

OK.


> I do not understand what you mean, I think I did not change logic anywhere, I just wrapped f and jacobian into the vector function. The case here is where single functions are stored ( $f = (f_1, f_2, ... f_n)$) hence the component wise access. If one function is stored, this is not done.


Ah, right, the single function one is below.